src/java.desktop/macosx/classes/sun/java2d/metal/MTLGraphicsConfig.java

```diff
@@ -0,0 +1,414 @@
+                    System.getProperty("java.home", "") + File.separator +
+                            "lib" + File.separator + "shaders.metallib");
+
+
+    private int pixfmt;
+    private BufferCapabilities bufferCaps;
+    private long pConfigInfo;
+    private ContextCapabilities mtlCaps;
+    private final MTLContext context;
+    private final Object disposerReferent = new Object();
+    private final int maxTextureSize;
+
+    private static native boolean isMetalFrameworkAvailable();
+    private static native boolean tryLoadMetalLibrary(int displayID, String shaderLib);
+    private static native long getMTLConfigInfo(int displayID, String mtlShadersLib);
+
+    /**
+     * Returns maximum texture size supported by Metal. Must be
+     * called under MTLRQ lock.
+     */
+    private static native int nativeGetMaxTextureSize();
+
+    static {
+        mtlAvailable = isMetalFrameworkAvailable();
+    }
+
+    private MTLGraphicsConfig(CGraphicsDevice device, int pixfmt,
+                              long configInfo, int maxTextureSize,
+                              ContextCapabilities mtlCaps) {
+        super(device);
+
+        this.pixfmt = pixfmt;
+        this.pConfigInfo = configInfo;
+        this.mtlCaps = mtlCaps;
+        this.maxTextureSize = maxTextureSize;
+        context = new MTLContext(MTLRenderQueue.getInstance());
+        // add a record to the Disposer so that we destroy the native
+        // MTLGraphicsConfigInfo data when this object goes away
+        Disposer.addRecord(disposerReferent,
+                new MTLGCDisposerRecord(pConfigInfo));
+    }
+
+    @Override
+    public Object getProxyKey() {
+        return this;
+    }
+
+    public SurfaceData createManagedSurface(int w, int h, int transparency) {
+        return MTLSurfaceData.createData(this, w, h,
+                getColorModel(transparency),
+                null,
+                MTLSurfaceData.TEXTURE);
+    }
+
+    public static MTLGraphicsConfig getConfig(CGraphicsDevice device,
+                                              int displayID, int pixfmt)
+    {
+        if (!mtlAvailable) {
+            return null;
+        }
+
+        if (!tryLoadMetalLibrary(displayID, mtlShadersLib)) {
+            return null;
+        }
+
+        long cfginfo = 0;
+        int textureSize = 0;
+        final String[] ids = new String[1];
+        MTLRenderQueue rq = MTLRenderQueue.getInstance();
+        rq.lock();
+        try {
+            // getMTLConfigInfo() creates and destroys temporary
+            // surfaces/contexts, so we should first invalidate the current
+            // Java-level context and flush the queue...
+            MTLContext.invalidateCurrentContext();
+            cfginfo = getMTLConfigInfo(displayID, mtlShadersLib);
+            if (cfginfo != 0L) {
+                textureSize = nativeGetMaxTextureSize();
+                // 7160609: GL still fails to create a square texture of this
```

Comment:
Is this bug valid for metal?

Reply:
This needs more verification. Because in case of metal as of now we return only one maxTextureSize value : 16384 and right now we are clamping it to half of it. Created JDK-8260644 for the same. But i will update these comments under JDK-8260640.

Reply:
(MTLGraphicsConfig.h)
// From "Metal Feature Set Tables"
// There are 2 GPU families for mac - MTLGPUFamilyMac1 and MTLGPUFamilyMac2
// Both of them support "Maximum 2D texture width and height" of 16384 pixels
// Note : there is no API to get this value, hence hardcoding by reading from the table